File: src/main/java/io/github/shaksternano/mediamanipulator/graphics/drawable/OutlinedTextDrawable.java
package io.github.shaksternano.mediamanipulator.graphics.drawable;

import io.github.shaksternano.mediamanipulator.image.util.ImageUtil;

import java.awt.*;
import java.awt.font.FontRenderContext;
import java.awt.font.GlyphVector;
import java.util.Objects;

public class OutlinedTextDrawable extends TextDrawable {

    private final Color TEXT_FILL_COLOR;
    private final Color TEXT_OUTLINE_COLOR;
    private final float TEXT_OUTLINE_WIDTH_RATIO;

    public OutlinedTextDrawable(String text, Color textFillColor, Color textOutlineColor, float textOutlineWidthRatio) {
        super(text);
        TEXT_FILL_COLOR = textFillColor;
        TEXT_OUTLINE_COLOR = textOutlineColor;
        TEXT_OUTLINE_WIDTH_RATIO = textOutlineWidthRatio;
    }

    @Override
    public void draw(Graphics2D graphics, int x, int y, long timestamp) {
        Font font = graphics.getFont();
        float textOutlineWidth = font.getSize2D() * TEXT_OUTLINE_WIDTH_RATIO;
        int actualX = (int) (x + textOutlineWidth);
        int actualY = y + graphics.getFontMetrics().getAscent();
        BasicStroke outlineStroke = new BasicStroke(textOutlineWidth);

        Color originalColor = graphics.getColor();
        Stroke originalStroke = graphics.getStroke();
        RenderingHints originalHints = graphics.getRenderingHints();

        Shape textShape = createTextShape(graphics);

        ImageUtil.configureTextDrawQuality(graphics);

        graphics.setColor(TEXT_OUTLINE_COLOR);
        graphics.setStroke(outlineStroke);
        graphics.translate(actualX, actualY);
        graphics.draw(textShape);

        graphics.setColor(TEXT_FILL_COLOR);
        graphics.fill(textShape);

        graphics.setColor(originalColor);
        graphics.setStroke(originalStroke);
        graphics.setRenderingHints(originalHints);
        graphics.translate(-actualX, -actualY);
    }

    @Override
    public int getWidth(Graphics2D graphicsContext) {
        Font font = graphicsContext.getFont();
        float textOutlineWidth = font.getSize2D() * TEXT_OUTLINE_WIDTH_RATIO;
        return (int) (createTextShape(graphicsContext).getBounds2D().getWidth() + textOutlineWidth * 2);
    }

    @Override
    public int getHeight(Graphics2D graphicsContext) {
        Font font = graphicsContext.getFont();
        float textOutlineWidth = font.getSize2D() * TEXT_OUTLINE_WIDTH_RATIO;
        return (int) (createTextShape(graphicsContext).getBounds2D().getHeight() + textOutlineWidth * 2);
    }

    private Shape createTextShape(Graphics2D graphicsContext) {
        Font font = graphicsContext.getFont();
        FontRenderContext fontRenderContext = graphicsContext.getFontRenderContext();
        GlyphVector glyphVector = font.createGlyphVector(fontRenderContext, getText());
        return glyphVector.getOutline();
    }

    @Override
    public Drawable resizeToWidth(int width) {
        throw new UnsupportedOperationException();
    }

    @Override
    public Drawable resizeToHeight(int height) {
        throw new UnsupportedOperationException();
    }

    @Override
    public int getFrameCount() {
        return 1;
    }

    @Override
    public boolean sameAsPreviousFrame() {
        return true;
    }

    @Override
    public int hashCode() {
        return Objects.hash(getText(), TEXT_FILL_COLOR, TEXT_OUTLINE_COLOR, TEXT_OUTLINE_WIDTH_RATIO);
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == this) {
            return true;
        } else if (obj instanceof OutlinedTextDrawable other) {
            return Objects.equals(getText(), other.getText())
                    && Objects.equals(TEXT_FILL_COLOR, other.TEXT_FILL_COLOR)
                    && Objects.equals(TEXT_OUTLINE_COLOR, other.TEXT_OUTLINE_COLOR)
                    && Objects.equals(TEXT_OUTLINE_WIDTH_RATIO, other.TEXT_OUTLINE_WIDTH_RATIO);
        } else {
            return false;
        }
    }

    @Override
    public String toString() {
        return getClass().getSimpleName() + "[Text: " + getText() + ", TextFillColor: " + TEXT_FILL_COLOR + ", TextOutlineColor: " + TEXT_OUTLINE_COLOR + ", TextOutlineWidth: " + TEXT_OUTLINE_WIDTH_RATIO + "]";
    }
}

File: core/src/main/kotlin/io/github/shaksternano/borgar/core/graphics/drawable/OutlinedTextDrawable.kt
package io.github.shaksternano.borgar.core.graphics.drawable

import io.github.shaksternano.borgar.core.graphics.DEFAULT_FONT_NAME
import io.github.shaksternano.borgar.core.graphics.bounds
import io.github.shaksternano.borgar.core.graphics.configureTextDrawQuality
import io.github.shaksternano.borgar.core.graphics.shape
import io.github.shaksternano.borgar.core.util.hash
import io.github.shaksternano.borgar.core.util.kClass
import java.awt.BasicStroke
import java.awt.Color
import java.awt.Font
import java.awt.Graphics2D
import kotlin.time.Duration

class OutlinedTextDrawable(
    override val text: String,
    private val fillColor: Color,
    private val outlineColor: Color,
    private val outlineWidthRatio: Double,
) : TextDrawable {

    override suspend fun draw(graphics: Graphics2D, x: Int, y: Int, timestamp: Duration) {
        val originalFont = graphics.font
        val originalColor = graphics.color
        val originalStroke = graphics.stroke
        val originalHints = graphics.renderingHints

        var x1 = x.toDouble()
        val splitText = splitTextFallbackFont(graphics)
        splitText.forEach { (substring, font) ->
            graphics.font = font
            val textOutlineWidth = font.size2D * outlineWidthRatio
            val actualX = x1 + textOutlineWidth
            val actualY = (y + graphics.fontMetrics.ascent).toDouble()
            val outlineStroke = BasicStroke(textOutlineWidth.toFloat())

            val textShape = graphics.shape(substring)

            graphics.configureTextDrawQuality()
            graphics.color = outlineColor
            graphics.stroke = outlineStroke
            graphics.translate(actualX, actualY)
            graphics.draw(textShape)

            graphics.color = fillColor
            graphics.fill(textShape)

            graphics.translate(-actualX, -actualY)
            x1 += graphics.bounds(substring).width + textOutlineWidth
        }

        graphics.font = originalFont
        graphics.color = originalColor
        graphics.stroke = originalStroke
        graphics.setRenderingHints(originalHints)
    }

    override fun getWidth(graphics: Graphics2D): Int {
        val originalFont = graphics.font
        val width = splitTextFallbackFont(graphics).sumOf { (substring, font) ->
            graphics.font = font
            val textOutlineWidth = font.size2D * outlineWidthRatio
            graphics.bounds(substring).width + textOutlineWidth * 2
        }
        graphics.font = originalFont
        return width.toInt()
    }

    override fun getHeight(graphics: Graphics2D): Int {
        val originalFont = graphics.font
        val height = splitTextFallbackFont(graphics).maxOf { (substring, font) ->
            graphics.font = font
            val textOutlineWidth = font.size2D * outlineWidthRatio
            graphics.bounds(substring).height + textOutlineWidth * 2
        }
        graphics.font = originalFont
        return height.toInt()
    }

    private fun splitTextFallbackFont(graphics: Graphics2D): List<Pair<String, Font>> {
        val mainFont = graphics.font
        val fallbackFont = Font(DEFAULT_FONT_NAME, Font.PLAIN, mainFont.size)
        val splitText = mutableListOf<Pair<String, Font>>()
        var substring = ""
        var currentFont = mainFont
        text.forEach {
            if (mainFont.canDisplay(it)) {
                if (currentFont == mainFont || substring.isEmpty()) {
                    substring += it
                } else {
                    splitText.add(substring to currentFont)
                    substring = it.toString()
                }
                currentFont = mainFont
            } else {
                if (currentFont == fallbackFont || substring.isEmpty()) {
                    substring += it
                } else {
                    splitText.add(substring to currentFont)
                    substring = it.toString()
                }
                currentFont = fallbackFont
            }
        }
        if (substring.isNotEmpty()) {
            splitText.add(substring to currentFont)
        }
        return splitText
    }

    override fun resizeToHeight(height: Int): Drawable? = null

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (kClass != other?.kClass) return false

        other as OutlinedTextDrawable

        if (text != other.text) return false
        if (fillColor != other.fillColor) return false
        if (outlineColor != other.outlineColor) return false
        if (outlineWidthRatio != other.outlineWidthRatio) return false

        return true
    }

    override fun hashCode(): Int = hash(
        text,
        fillColor,
        outlineColor,
        outlineWidthRatio,
    )

    override fun toString(): String {
        return "OutlinedTextDrawable(text='$text'" +
            ", fillColor=$fillColor" +
            ", outlineColor=$outlineColor" +
            ", outlineWidthRatio=$outlineWidthRatio)"
    }
}
